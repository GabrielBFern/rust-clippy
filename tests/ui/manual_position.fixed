#![warn(clippy::manual_position)]
#![allow(clippy::needless_return)]

fn main() {
    unwrap_test();
    expect_test();
    try_test();
    map_test();
    find_test();
    deref_test();
    rposition_test();
}

fn expect_test() {
    let v = [1, 2, 3, 5, 6];
    let _ = v.iter().position(|&ch|ch == 3).expect("");
    let _ = v.iter().enumerate().find(|(_, &ch)| ch == 3).expect("").1;
}

fn unwrap_test() {
    let v = [1, 2, 3, 5, 6];
    let _ = v.iter().position(|&ch|ch == 3).unwrap();
    let _ = v.iter().enumerate().find(|(_, &ch)| ch == 3).unwrap().1;
}

fn try_test() -> Option<()> {
    let v = [1, 2, 3, 5, 6];
    let _ = v.iter().position(|&ch|ch == 3)?;
    let _ = v.iter().enumerate().find(|(_, &ch)| ch == 3)?.1;
    Some(())
}

fn map_test() {
    let v = [1, 2, 3, 5, 6];
    let _ = v.iter().position(|&ch|ch == 3);
    #[rustfmt::skip]
    let _ = v.iter().position(|&ch|ch == 3);
    let _ = v.iter().position(|&ch|ch == 3);
    let _ = v.iter().position(|&ch|ch == 3);

    // Only trigger when returning the first field of the result
    let x = (1, 2);
    let _ = v.iter().enumerate().find(|(_, &ch)| ch == 3).map(|_| x.0);
}

fn find_test() {
    let v = [1, 2, 3, 5, 6];
    let _ = v.iter().position(|&ch|ch == 3).unwrap();
    // Doesn't trigger when the predicate use the index
    let _ = v.iter().enumerate().find(|&(i, _)| i >= 1).unwrap().0;
    let _ = v.iter().enumerate().find(|&(i, &ch)| ch == 3 && i >= 1).unwrap().0;
}

fn deref_test() {
    let v = [1, 2, 3, 5, 6];
    let _ = v.iter().position(|&ch|ch == 3).unwrap();
    let _ = v.iter().position(|&ch|ch == 3).unwrap();
    let _ = v.iter().position(|ch|*ch == 3).unwrap();
    // let _ = v.iter().enumerate().find(|(_, ch)| **ch == 3).unwrap().0;
    // Sugestion make:
    // let _ = v.iter().position(|ch|**ch == 3).unwrap();
    // but need to be:
    // let _ = v.iter().position(|ch|*ch == 3).unwrap();
    // or
    // let _ = v.iter().position(|&ch|ch == 3).unwrap();
}

fn rposition_test() {
    let v = [1, 2, 3, 5, 3, 6];
    let _ = v.iter().rposition(|&ch|ch == 3).unwrap();
}
