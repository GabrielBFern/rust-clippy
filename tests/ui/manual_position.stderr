error: manual implementation of position
  --> tests/ui/manual_position.rs:16:22
   |
LL |     let _ = v.iter().enumerate().find(|(_, &ch)| ch == 3).expect("").0;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `position(|&ch|ch == 3).expect("")`
   |
   = note: `-D clippy::manual-position` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::manual_position)]`

error: manual implementation of position
  --> tests/ui/manual_position.rs:22:22
   |
LL |     let _ = v.iter().enumerate().find(|(_, &ch)| ch == 3).unwrap().0;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `position(|&ch|ch == 3).unwrap()`

error: manual implementation of position
  --> tests/ui/manual_position.rs:28:22
   |
LL |     let _ = v.iter().enumerate().find(|(_, &ch)| ch == 3)?.0;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `position(|&ch|ch == 3)?`

error: manual implementation of position
  --> tests/ui/manual_position.rs:35:22
   |
LL |     let _ = v.iter().enumerate().find(|&(_, &ch)| ch == 3).map(|c| c.0);
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `position(|&ch|ch == 3)`

error: manual implementation of position
  --> tests/ui/manual_position.rs:37:22
   |
LL |     let _ = v.iter().enumerate().find(|(_, &ch)| ch == 3).map(|c| { return c.0  });
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `position(|&ch|ch == 3)`

error: manual implementation of position
  --> tests/ui/manual_position.rs:38:22
   |
LL |     let _ = v.iter().enumerate().find(|(_, &ch)| ch == 3).map(|(p, _)| p);
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `position(|&ch|ch == 3)`

error: manual implementation of position
  --> tests/ui/manual_position.rs:39:22
   |
LL |       let _ = v.iter().enumerate().find(|(_, &ch)| ch == 3).map(|(p, ..)| {
   |  ______________________^
LL | |         return p;
LL | |     });
   | |______^ help: replace with: `position(|&ch|ch == 3)`

error: manual implementation of position
  --> tests/ui/manual_position.rs:50:22
   |
LL |     let _ = v.iter().enumerate().find(|(_, &ch)| ch == 3).unwrap().0;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `position(|&ch|ch == 3).unwrap()`

error: manual implementation of position
  --> tests/ui/manual_position.rs:58:22
   |
LL |     let _ = v.iter().enumerate().find(|(_, &ch)| ch == 3).unwrap().0;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `position(|&ch|ch == 3).unwrap()`

error: manual implementation of position
  --> tests/ui/manual_position.rs:59:22
   |
LL |     let _ = v.iter().enumerate().find(|&(_, &ch)| ch == 3).unwrap().0;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `position(|&ch|ch == 3).unwrap()`

error: manual implementation of position
  --> tests/ui/manual_position.rs:60:22
   |
LL |     let _ = v.iter().enumerate().find(|&(_, ch)| *ch == 3).unwrap().0;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `position(|ch|*ch == 3).unwrap()`

error: manual implementation of rposition
  --> tests/ui/manual_position.rs:72:22
   |
LL |     let _ = v.iter().enumerate().rev().find(|(_, &ch)| ch == 3).unwrap().0;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `rposition(|&ch|ch == 3).unwrap()`

error: aborting due to 12 previous errors

